crates/uv-distribution/src/workspace.rs

```diff
@@ -824,6 +808,55 @@ fn is_excluded_from_workspace(
     Ok(false)
 }
 
+/// An iterator over a [`Pattern`] in a workspace.
+///
+/// Iterates from the workspace root, returning paths that match the pattern.
+#[derive(Debug)]
+struct WorkspaceGlobIter<'a> {
+    root: &'a Path,
+    glob: &'a Pattern,
+    iter: walkdir::IntoIter,
+}
+
+impl<'a> WorkspaceGlobIter<'a> {
+    fn new(root: &'a Path, glob: &'a Pattern) -> Self {
+        Self {
+            root,
+            glob,
+            iter: walkdir::WalkDir::new(root).into_iter(),
+        }
+    }
+}
+
+impl<'a> Iterator for WorkspaceGlobIter<'a> {
+    type Item = Result<PathBuf, walkdir::Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let Some(entry) = self.iter.next() else {
+                return None;
+            };
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => return Some(Err(err)),
+            };
+            if entry.file_type().is_dir() {
+                if self.glob.matches_path(entry.path())
+                    || entry
+                        .path()
+                        .strip_prefix(self.root)
+                        .is_ok_and(|path| self.glob.matches_path(path))
+                {
+                    // Match, but avoid recursing into the directory.
+                    let path = entry.path().to_path_buf();
+                    self.iter.skip_current_dir();
```

Comment:
I have a vague feeling that this isn't correct for some recursive patterns? Like if you had `members = ["crates/**"]`, then we'd only return the top-level directory? But that might be nonsensical for workspaces anyway?

I'm not sure.

But if I don't do this, then we discover files like `crates/uv/src` in addition to `crates/uv`, given `members = ["crates/*"]`.

Reply:
Hmmm. Should there be more detection logic than just "matches glob"? So yeah, this iterator would return `crates/uv/src`, but if `crates/uv/src` doesn't have a `Cargo.toml` in it, then it can't be a package directory right? I'm not sure if that filtering logic should be here or not, but it seems like it's definitely wrong to skip recursion here.

Reply:
We do check that elsewhere, but I think we require that any discovered directories include a `Cargo.toml`, rather than _using_ `Cargo.toml` to determine if it's a package. Cargo has the same behavior. E.g., changing uv to use `members = ["crates/*", "python/*"]` gives:

```
❯ cargo check
error: failed to load manifest for workspace member `/Users/crmarsh/workspace/uv/python/uv`
referenced by workspace at `/Users/crmarsh/workspace/uv/Cargo.toml`

Caused by:
  failed to read `/Users/crmarsh/workspace/uv/python/uv/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
```

Reply:
If I could check whether the glob was recursive, this would be easy. `glob` does that, but it doesn't expose `is_recursive` on `glob`.

Reply:
I guess the other option (implemented here) is `require_literal_separator` which means `members = ["crates/*"]` only matches at the top-level anyway.

Reply:
Yeah setting `require_literal_separator` is probably a good idea anyway, especially when recursive globs are supported.

Reply:
IDK I'm really torn on what the right behavior is here. I feel like I'm unnecessarily recursing in the common case. Glob has:
```rust
if idx == self.dir_patterns.len() - 1 {
    // it is not possible for a pattern to match a directory
    // *AND* its children so we don't need to check the
    // children

    if !self.require_dir || is_dir(&path) {
        return Some(Ok(path));
    }
}
```

But that only applies if they don't have `self.dir_patterns[idx].is_recursive`? Maybe this whole change is dumb.